File: fc-chat/demo/part3/ch3/05-1-message-system-add-channel/src/main/java/net/prostars/messagesystem/entity/UserChannelEntity.java
package net.prostars.messagesystem.entity;

import jakarta.persistence.*;
import java.util.Objects;

@Entity
@Table(name = "user_channel")
@IdClass(UserChannelId.class)
public class UserChannelEntity extends BaseEntity {

  @Id
  @Column(name = "user_id", nullable = false)
  private Long userId;

  @Id
  @Column(name = "channel_id", nullable = false)
  private Long channelId;

  @Column(name = "last_read_msg_seq", nullable = false)
  private long lastReadMsgSeq;

  public UserChannelEntity() {}

  public UserChannelEntity(Long userId, Long channelId, long lastReadMsgSeq) {
    this.userId = userId;
    this.channelId = channelId;
    this.lastReadMsgSeq = lastReadMsgSeq;
  }

  public Long getUserId() {
    return userId;
  }

  public Long getChannelId() {
    return channelId;
  }

  public long getLastReadMsgSeq() {
    return lastReadMsgSeq;
  }

  public void setLastReadMsgSeq(long lastReadMsgSeq) {
    this.lastReadMsgSeq = lastReadMsgSeq;
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || getClass() != o.getClass()) return false;
    UserChannelEntity that = (UserChannelEntity) o;
    return Objects.equals(userId, that.userId) && Objects.equals(channelId, that.channelId);
  }

  @Override
  public int hashCode() {
    return Objects.hash(userId, channelId);
  }

  @Override
  public String toString() {
    return "UserChannelEntity{userId=%d, channelId=%d, lastReadMsgSeq=%d}"
        .formatted(userId, channelId, lastReadMsgSeq);
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/entity/UserChannelEntity.kt
package com.yscorp.withpush.messagesystem.entity

import jakarta.persistence.*
import java.util.*


@Entity
@Table(name = "user_channel")
@IdClass(UserChannelId::class)
class UserChannelEntity(

    @Id
    @Column(name = "user_id", nullable = false)
    var userId: Long? = null,

    @Id
    @Column(name = "channel_id", nullable = false)
    var channelId: Long? = null,

    @Column(name = "last_read_msg_seq", nullable = false)
    var lastReadMsgSeq: Long = 0,

    ) : BaseEntity() {


    override fun equals(o: Any?): Boolean {
        if (o == null || javaClass != o.javaClass) return false
        val that = o as UserChannelEntity
        return userId == that.userId && channelId == that.channelId
    }

    override fun hashCode(): Int {
        return Objects.hash(userId, channelId)
    }

    override fun toString(): String {
        return "UserChannelEntity{userId=%d, channelId=%d, lastReadMsgSeq=%d}"
            .formatted(userId, channelId, lastReadMsgSeq)
    }
}
